File: Difficulty/Medium/[36] Valid Sudoku/solution.py
#
# @lc app=leetcode id=36 lang=python3
#
# [36] Valid Sudoku
#
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        _dict = dict()
        for i in range(9):
            for j in range(9):
                if board[i][j] != '.':
                    b = '(' + board[i][j] + ')'
                    if _dict.get(str(i) + b, 0) or \
                       _dict.get(b + str(j), 0) or \
                       _dict.get(str(i // 3) + b + str(j // 3), 0):
                        return False
                    else:
                        _dict[str(i) + b] = 1
                        _dict[b + str(j)] = 1
                        _dict[str(i // 3) + b + str(j // 3)] = 1
        return True
        

File: Difficulty/Medium/[36] Valid Sudoku/solution.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        unordered_map<string, int> _map;
        for (int i = 0; i < 9; ++i){
            for (int j = 0; j < 9; ++j){
                if (board[i][j] != '.'){
                    string b = '.' + to_string(board[i][j]) + '.';
                    if (_map.find(to_string(i) + b) != _map.end() || 
                        _map.find(b + to_string(j)) != _map.end() || 
                        _map.find(to_string(i / 3) + b + to_string(j / 3)) != _map.end())
                        return false;
                    else{
                        _map.insert({to_string(i) + b, 1});
                        _map.insert({b + to_string(j), 1});
                        _map.insert({to_string(i / 3) + b + to_string(j / 3), 1});
                    }
                }
            }
        }
        return true;
    }
};